docs/training/apis-howto/api-core-ug.rst

```diff
@@ -1,212 +1,347 @@
+ability to stop and restart training. In this section, we’ll modify our script for the purpose of
+checkpointing.
+
+.. note::
+
+   Two Types of Continuations (Stopping and Restarting)
+
+   There are two types of continuations (methods for stopping and restarting training): pausing and
+   reactivating training using the WebUI or clicking **Continue Trial** after the experiment
+   completes.
+
+   Each type of continuation has its own behavior. While you always want to preserve the value you
+   are incrementing (the “model weight”), you do not always want to preserve the batch index. When
+   you pause and reactivate you want training to continue from the same batch index. However, when
+   starting a fresh experiment, you want training to start with a fresh batch index.
+
+   To distinguish between the two types of continuations, you can save the trial ID in the
+   checkpoint.
+
+In this step, we’ll run our experiment using the ``model_def_checkpoints.py`` script and its
+accompanying ``checkpoints.yaml`` experiment configuration file.
+
+Step 3.1: Save Checkpoints
+==========================
+
+To save checkpoints, add the ``store_path`` function to your script:
+
+.. code:: python
+
+   # NEW: Save checkpoint.
+   checkpoint_metadata_dict = {"steps_completed": steps_completed}
+
+   # NEW: Here we are saving multiple files to our checkpoint directory.
+   # 1) a model state file and 2) a file includes information
+   # about the training loop state.
+   with core_context.checkpoint.store_path(checkpoint_metadata_dict) as (path, storage_id):
+         torch.save(model.state_dict(), path / "checkpoint.pt")
+         with path.joinpath("state").open("w") as f:
+             f.write(f"{epochs_completed},{info.trial.trial_id}")
+
+Step 3.2: Enable Pausing and Resuming an Experiment
+===================================================
+
+To enable pausing an experiment, enable preemption:
+
+.. code:: python
+
+   # NEW: Detect when the experiment is paused by the WebUI.
+   if core_context.preempt.should_preempt():
+       return
+
+define a load function:
+
+.. code:: python
+
+   # NEW: Define load_state function for restarting model training from existing checkpoint. Returns (.pt, int).
+   # Also update load_state header to take trial info object as an argument.
+   def load_state(checkpoint_directory, trial_id):
+       checkpoint_directory = pathlib.Path(checkpoint_directory)
+
+       with checkpoint_directory.joinpath("checkpoint.pt").open("rb") as f:
+           model = torch.load(f)
+
+       with checkpoint_directory.joinpath("state").open("r") as f:
+           epochs_completed, ckpt_trial_id = [int(field) for field in f.read().split(",")]
+
+.. code:: python
+
+   # If trial ID does not match our current trial ID, we'll ignore epochs
+   # completed and start training from epoch_idx = 0
+   if ckpt_trial_id != trial_id:
+       epochs_completed = 0
+
+   return model, epochs_completed
```

Comment:
this is all supposed to be one function, right?  I think it should be one code block, not two separate ones.

Reply:
model_def_checkpoints.py has since been modified and this code block has changed; resolved